execution/block-partitioner/src/sharded_block_partitioner/conflict_detector.rs

```diff
@@ -0,0 +1,196 @@
+    ) -> (
+        Vec<AnalyzedTransaction>,
+        Vec<CrossShardDependencies>,
+        Vec<AnalyzedTransaction>,
+    ) {
+        // Iterate through all the transactions and if any shard has taken read/write lock on the storage location
+        // and has a higher priority than this shard id, then this transaction needs to be moved to the end of the block.
+        let mut accepted_txns = Vec::new();
+        let mut accepted_txn_dependencies = Vec::new();
+        let mut rejected_txns = Vec::new();
+        for (_, txn) in txns.into_iter().enumerate() {
+            if self.check_for_cross_shard_conflict(self.shard_id, &txn, cross_shard_rw_set) {
+                rejected_txns.push(txn);
+            } else {
+                accepted_txn_dependencies.push(self.get_dependencies_for_frozen_txn(
+                    &txn,
+                    Arc::new(vec![]),
+                    prev_rounds_rw_set_with_index.clone(),
+                ));
+                accepted_txns.push(txn);
+            }
+        }
+        (accepted_txns, accepted_txn_dependencies, rejected_txns)
+    }
+
+    /// Adds a cross shard dependency for a transaction. This can be done by finding the maximum transaction index
+    /// that has taken a read/write lock on the storage location the current transaction is trying to read/write.
+    /// We traverse the current round read/write set in reverse order starting from shard id -1 and look for the first
+    /// txn index that has taken a read/write lock on the storage location. If we can't find any such txn index, we
+    /// traverse the previous rounds read/write set in reverse order and look for the first txn index that has taken
+    /// a read/write lock on the storage location.
+    fn get_dependencies_for_frozen_txn(
+        &self,
+        frozen_txn: &AnalyzedTransaction,
+        current_round_rw_set_with_index: Arc<Vec<RWSetWithTxnIndex>>,
+        prev_rounds_rw_set_with_index: Arc<Vec<RWSetWithTxnIndex>>,
+    ) -> CrossShardDependencies {
+        if current_round_rw_set_with_index.is_empty() && prev_rounds_rw_set_with_index.is_empty() {
+            return CrossShardDependencies::default();
+        }
+        // Iterate through the frozen dependencies and add the max transaction index for each storage location
+        let mut cross_shard_dependencies = CrossShardDependencies::default();
+        for read_location in frozen_txn.read_set().iter() {
+            // For current round, iterate through all shards less than current shards in the reverse order and for previous rounds iterate through all shards in the reverse order
+            // and find the first shard id that has taken a write lock on the storage location. This ensures that we find the highest txn index that is conflicting
+            // with the current transaction.
+            for rw_set_with_index in current_round_rw_set_with_index
+                .iter()
+                .take(self.shard_id)
```

Comment:
missing a `.rev()`? I didn't find `current_round_rw_set_with_index` was reversed anywhere else...

Reply:
Good eye, fixed it.